wallet/test/Test/Pos/Wallet/Web/Util.hs

```diff
@@ -0,0 +1,210 @@
+import           Pos.Context                    (LastKnownHeaderTag, ProgressHeaderTag)
+import           Pos.Core                       (Address, BlockCount, Coin,
+                                                 HasConfiguration, generatedSecrets,
+                                                 gsSecretKeysPoor, headerHashG)
+import           Pos.Core.Address               (IsBootstrapEraAddr (..),
+                                                 deriveLvl2KeyPair)
+import           Pos.Crypto                     (EncryptedSecretKey, PassPhrase,
+                                                 ShouldCheckPassphrase (..),
+                                                 firstHardened)
+import           Pos.Generator.Block            (genBlocks)
+import           Pos.Launcher                   (HasConfigurations)
+import           Pos.StateLock                  (Priority (..), modifyStateLock)
+import           Pos.Txp.Core                   (TxIn, TxOut (..), TxOutAux (..))
+import           Pos.Txp.Toil                   (Utxo)
+import           Pos.Util.Chrono                (OldestFirst (..))
+import           Pos.Util.CompileInfo           (HasCompileInfo)
+import           Pos.Util.Servant               (encodeCType)
+import           Pos.Util.UserSecret            (mkGenesisWalletUserSecret)
+import           Pos.Util.Util                  (HasLens (..), _neLast)
+import           Pos.Wallet.Web.ClientTypes     (Addr, CId)
+import           Pos.Wallet.Web.Methods.Restore (importWalletDo)
+
+import           Test.Pos.Block.Logic.Util      (EnableTxPayload, InplaceDB,
+                                                 genBlockGenParams)
+import           Test.Pos.Util                  (assertProperty, maybeStopProperty)
+import           Test.Pos.Wallet.Web.Mode       (WalletProperty)
+
+----------------------------------------------------------------------------
+-- Block utils
+----------------------------------------------------------------------------
+
+-- | Gen blocks in WalletProperty
+wpGenBlocks
+    :: (HasCompileInfo, HasConfigurations)
+    => Maybe BlockCount
+    -> EnableTxPayload
+    -> InplaceDB
+    -> WalletProperty (OldestFirst [] Blund)
+wpGenBlocks blkCnt enTxPayload inplaceDB = do
+    params <- genBlockGenParams blkCnt enTxPayload inplaceDB
+    g <- pick $ MkGen $ \qc _ -> qc
+    lift $ modifyStateLock HighPriority "wpGenBlocks" $ \prevTip -> do
+        blunds <- evalRandT (genBlocks params) g
+        case nonEmpty $ getOldestFirst blunds of
+            Just nonEmptyBlunds -> do
+                let tipBlockHeader = nonEmptyBlunds ^. _neLast . _1 . blockHeader
+                lastKnownHeader <- view (lensOf @LastKnownHeaderTag)
+                atomically $ writeTVar lastKnownHeader (Just tipBlockHeader)
+                progressHeader <- view (lensOf @ProgressHeaderTag)
+                atomically $ do
+                    void $ tryTakeTMVar progressHeader
```

Comment:
How does this work?

Reply:
we take away header if it's presented in `TMVar` and put new one

Reply:
Ah rly, thanks